cpp/src/spatial/hausdorff.cu

```diff
@@ -121,38 +121,41 @@ struct hausdorff_functor {
 
     auto scatter_out = make_scatter_output_iterator(result_temp.begin(), scatter_map);
 
-    auto gpc_key_iter = thrust::make_transform_iterator(
+    auto gcp_key_iter = thrust::make_transform_iterator(
       gcp_iter, [] __device__(cartesian_product_group_index const& idx) {
         return thrust::make_pair(idx.group_a.idx, idx.group_b.idx);
       });
 
     // the following output iterator and `inclusive_scan_by_key` could be replaced by a
-    // reduce_by_key, if it supported non-commutative operators.
+    // reduce_by_key, if it supported non-commutative operators and more inputs (int 64).
 
-    auto const num_cartesian =
-      static_cast<uint64_t>(num_points) * static_cast<uint64_t>(num_points);
+    // copy offsets to host for batching purposes.
 
-    //
-    // `thrust::inclusive_scan_by_key` causes an out-of-memory
-    // error on input sizes close to (but not exactly) INT_MAX.
-    //
-    // Doing the reduction in chunks is a temporary workaround until this is fixed.
-    //
-    // The magic number between OOM vs. no OOM is somewhere between:
-    //                                             (1uL << 31uL) - 2048uL;
-    auto const magic_inclusive_scan_by_key_limit = (1uL << 31uL) - 4096uL;
+    thrust::host_vector<uint32_t> h_space_offsets(space_offsets.size());
 
-    for (auto itr = gpc_key_iter, end = gpc_key_iter + num_cartesian; itr < end;) {
-      auto const len = static_cast<uint32_t>(std::min(
-        static_cast<uint64_t>(thrust::distance(itr, end)), magic_inclusive_scan_by_key_limit));
+    CUDA_TRY(cudaMemcpyAsync(h_space_offsets.data(),
+                             space_offsets.data<uint32_t>(),
+                             space_offsets.size() * sizeof(uint32_t),
+                             cudaMemcpyDeviceToHost,
+                             stream.value()));
+
+    stream.synchronize();
+
+    for (uint32_t i = 1; i <= h_space_offsets.size(); i++) {
```

Comment:
What does this raw loop do? Please add comments. Also, since the only thing indexed by `i` and `i-1` is `h_space_offsets`, seems like you could  use an algorithm here (`adjacent_something`?)